gothemis/session/session.go

```diff
@@ -37,10 +37,23 @@ const (
 
 // Errors returned by Secure Session.
 var (
-	ErrMissingClientID   = errors.NewWithCode(errors.InvalidParameter, "empty client ID for Secure Session")
-	ErrMissingPrivateKey = errors.NewWithCode(errors.InvalidParameter, "empty client private key for Secure Session")
-	ErrMissingMessage    = errors.NewWithCode(errors.InvalidParameter, "empty message for Secure Session")
-	ErrOverflow          = errors.NewWithCode(errors.NoMemory, "Secure Session cannot allocate enough memory")
+	ErrDestroySessionObject      = errors.New("Failed to destroy secure session object")
+	ErrCreateSessionObject       = errors.New("Failed to create secure session object")
+	ErrGetRequestSize            = errors.New("Failed to get request size")
+	ErrGenRequest                = errors.New("Failed to generate request")
+	ErrGetWrappedSize            = errors.New("Failed to get wrapped size")
+	ErrWrapData                  = errors.New("Failed to wrap data")
+	ErrGetUnwrappedSize          = errors.New("Failed to get unwrapped size")
+	ErrUnwrapData                = errors.New("Failed to unwrap data")
+	ErrCallbackGetUnwrappedSize  = errors.NewCallbackError("Failed to get unwraped size (get_public_key_by_id callback error)")
+	ErrCallbackUnwrapData        = errors.NewCallbackError("Failed to unwrap data (get_public_key_by_id callback error)")
+	ErrCallbackGetRemoteIDLen    = errors.NewCallbackError("Failed to get session remote id length")
+	ErrCallbackBadRemoteIDLen    = errors.NewCallbackError("Incorrect remote id length (0)")
+	ErrCallbackGetRemoteID       = errors.NewCallbackError("Failed to get session remote id")
```

Comment:
Honestly, I'd just squash all those variations into just two:

- `ErrMessageSize` — when we fail to get appropriate message size
- `ErrMessageData` — when processing something goes wrong

There may be some interest in whether the first or the second call to Themis Core failed, but I don't think that we need 10 new API entries to express basically the same thing: your data seems to be broken.

What do you think, everyone?

Reply:
agree

Reply:
actually, we had some situations when we read logs and messages like that which include context where it occurs, could save our time) actually in Acra, where we can meet errors in keystore, which use secure cell, in transport package which uses secure session, in decryption module, which uses secure cell + secure message. And info about where it occurs, like could not process something in cell or message, or session, could help to understand part where was error.

Reply:
In my opinion, there isn't really much use for having an error for each individual method.

Themis Core has a signle THEMIS_FAIL for all failures, not a bunch of

- THEMIS_SECURE_CELL_SEAL_ENCRYPT_FAIL
- THEMIS_SECURE_CELL_TOKEN_PROTECT_DECRYPT_FAIL
- THEMIS_SECURE_MESSAGE_VERIFY_FAIL

unique for each method. The context should be provided by a higher level means.

Discriminating between Secure Cell and Secure Message failures may be useful for a system where Secure Cell is used in only one subsystem and Secure Message is used in a different one. Then by observing the messages you can guess where the failure is. However, if multiple subsystems use Secure Cell then you still need some context for disambiguation – context added by the application, not Themis. And if you have this context in the first place, you don't really need to rely on Cell/Message distinction specifically to pinpoint the faulty subsystem.